### PR TITLE
add load 3d node support

### PR DESCRIPTION
**Pull request by jtydhr88**
Hi,
I implemented a node to load local 3d models and output IMAGE for furthur process, such as controlnet, as reference image.
A test workflow 

[load-3d (1).json](https://github.com/user-attachments/files/17710614/load-3d.1.json)

screenshot
image

Need https://github.com/comfyanonymous/ComfyUI/pull/5564 as BE

### CODE REVIEW

**Review by huchenlei — COMMENTED** [review]
Wow +27MB on package size. This threejs lib is huuuuge. If it is as simple as viewing 3D object without requiring fancy features like particles and shaders, can we use a more lightweight library instead? Or is there plan to add more features here?

**Comment by huchenlei** [comment]
Update:

threejs is probably fine as the binary size of dist only goes from 3.55 MB to 4.36 MB.

File: public/assets/sorted-custom-node-map.json [hunk 1]
```diff
@@ -2599,5 +2598,6 @@
   "Perlin Power Fractal Latent (PPF Noise)": 1,
   "Cross-Hatch Power Fractal (PPF Noise)": 1,
   "Linear Cross-Hatch Power Fractal (PPF Noise)": 1,
-  "CLIPTextEncodeControlnet": 1
+  "CLIPTextEncodeControlnet": 1,
+  "Load3D": 1
```
Comment (huchenlei) [inline public/assets/sorted-custom-node-map.json]: You don't need to update this list as it is automatically generated, and not supposed to be always up-to-date.

Comment (jtydhr88) [reply public/assets/sorted-custom-node-map.json]: removed

**Comment by jtydhr88** [comment]
image
use comfy native components instead

**Review by huchenlei — CHANGES_REQUESTED** [review]
Please fix:
```
Error: src/extensions/core/load3d.ts(528,14): error TS2339: Property 'name' does not exist on type 'LGraphNode'.
Error: src/extensions/core/load3d.ts(540,14): error TS2339: Property 'load3d' does not exist on type 'LGraphNode'.
Error: src/extensions/core/load3d.ts(540,34): error TS2345: Argument of type 'LGraphNode' is not assignable to parameter of type 'objectOutputType<{ id: ZodUnion<[ZodNumber, ZodString]>; type: ZodString; pos: ZodUnion<[ZodEffects<ZodObject<{ 0: ZodNumber; 1: ZodNumber; }, "passthrough", ZodTypeAny, objectOutputType<...>, objectInputType<...>>, number[], objectInputType<...>>, ZodTuple<...>]>; ... 9 more ...; bgcolor: ZodOptional<...>; }, ZodTy...'.
  Type 'LGraphNode' is not assignable to type '{ color?: string; size?: number[] | [number, number, ...unknown[]]; mode?: number; id?: string | number; type?: string; pos?: number[] | [number, number, ...unknown[]]; flags?: { collapsed?: boolean; pinned?: boolean; allow_interaction?: boolean; horizontal?: boolean; skip_repeated_outputs?: boolean; } & { ...; }; ....'.
    Types of property 'size' are incompatible.
      Type 'Size' is not assignable to type 'number[] | [number, number, ...unknown[]]'.
        Type 'Float32Array' is not assignable to type 'number[] | [number, number, ...unknown[]]'.
          Type 'Float32Array' is not assignable to type '[number, number, ...unknown[]]'.
Error: src/extensions/core/load3d.ts(548,20): error TS2339: Property 'load3d' does not exist on type 'LGraphNode'.
Error: src/extensions/core/load3d.ts(549,18): error TS2339: Property 'load3d' does not exist on type 'LGraphNode'.
Error: src/extensions/core/load3d.ts(556,20): error TS2339: Property 'load3d' does not exist on type 'LGraphNode'.
Error: src/extensions/core/load3d.ts(557,18): error TS2339: Property 'load3d' does not exist on type 'LGraphNode'.
Error: src/extensions/core/load3d.ts(564,16): error TS2339: Property 'load3d' does not exist on type 'LGraphNode'.
```

You can run `npm run typecheck` locally to verify things are compilable.

**Comment by jtydhr88** [comment]
@huchenlei I don't think I should add any properties to LGraphNode, so I add ts-strict-ignore back for now

File: src/extensions/core/load3d.ts [hunk 2]
```diff
@@ -0,0 +1,735 @@
...
+
+        node.load3d = new Load3d(node, container)
+
+        node.onResize = function () {
+          let [w, h] = this.size
+          if (w <= 300) w = 300
+          if (h <= 300) h = 300
+          this.size = [w, h]
+
+          if (this.load3d) {
```
Comment (huchenlei) [inline src/extensions/core/load3d.ts]: `this` ref is ambiguous here. Please replace it with `node` if you are trying to bind something to the LGraphNode object. With previous comment in mind, here you probably want to read from widget object instead.

File: src/extensions/core/load3d.ts [hunk 3]
```diff
@@ -0,0 +1,735 @@
...
+
+        containerWrapper.id = `load-3d-wrapper-${inputName.toLowerCase()}`
+        containerWrapper.classList.add('comfyui-load-3d-wrapper')
+
+        const container = document.createElement('div')
+        container.id = `load-3d-${inputName.toLowerCase()}`
+        container.classList.add('comfyui-load-3d')
+        containerWrapper.appendChild(container)
+
+        node.load3d = new Load3d(node, container)
```
Comment (huchenlei) [inline src/extensions/core/load3d.ts]: I don't think you are supposed to bind the `load3d` to LGraphNode, as a node can have multiple widget of same type in theory. If a backend node specifies multiple Load3D widgets, this impl will cause issue. Let's keep it local to the widget.

**Review by huchenlei — CHANGES_REQUESTED** [review]
Please enable ts-strict. The errors are actually meaningful. Also current errors are from non-strict tsc. 😞

File: src/extensions/core/load3d.ts (hunk 2 above)
Comment (jtydhr88) [reply src/extensions/core/load3d.ts]: fixed

File: src/extensions/core/load3d.ts (hunk 3 above)
Comment (jtydhr88) [reply src/extensions/core/load3d.ts]: fixed

**Comment by jtydhr88** [comment]
Still have a few things need to correct, I will fix today

**Comment by jtydhr88** [comment]
please help to run the tests @huchenlei

**Comment by jtydhr88** [comment]
close it for now, I will send a new PR

**Comment by jtydhr88** [comment]
please review https://github.com/Comfy-Org/ComfyUI_frontend/pull/1563 instead

**Pull request closed without merging**